cloudant-client/src/main/java/com/cloudant/client/api/query/QueryBuilder.java

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright © 2017, 2018 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.cloudant.client.api.query;
+
+import com.cloudant.client.internal.query.Helpers;
+import com.cloudant.http.HttpConnection;
+
+import java.util.LinkedList;
+
+/**
+ * <p>
+ *     Helper class for building query selector strings.
+ * </p>
+ * <p>
+ *     Example usage to return the name and year of movies starring Alec Guinness since 1960
+ *     with the results sorted by year descending:
+ * </p>
+ * <pre>
+ * {@code
+ * String query = db.query(new QueryBuilder(and(
+ *   gt("Movie_year", 1960),
+ *   eq("Person_name", "Alec Guinness")).toString()).
+ *   sort(Sort.desc("Movie_year")).
+ *   fields("Movie_name", "Movie_year").
+ *   build();
+ * }
+ * </pre>
+ * <p>
+ *     The resulting string can then be used to query the database using either
+ *     {@link com.cloudant.client.api.Database#query(String, Class)} (which deserialises the results)
+ *     or {@link com.cloudant.client.api.CloudantClient#executeRequest(HttpConnection)} (which can
+ *     be used to make a "raw" request by POSTing the string to the {@code /_find} endpoint).
+ * </p>
+ * @see Selector
+ */
+public class QueryBuilder {
+
+    private final String selector;
+    private String[] fields;
+    private Sort[] sort;
+    private Long limit;
+    private Long skip;
+    private String bookmark;
+    private boolean update = true;
+    private boolean stable;
+    private boolean executionStats;
+    private String[] useIndex;
+
+    /**
+     * <p>
+     * Construct a {@code QueryBuilder} object for use with
+     * {@link com.cloudant.client.api.Database#query(String, Class)}
+     * </p>
+     * <p>
+     * The easiest way of obtaining a selector string is to obtain a {@link Selector} from an
+     * {@link Operation} or {@link Expression} and call {@code toString} on the resulting
+     * {@link Selector}.
+     * </p>
+     * @param selector string representation of a JSON object describing criteria used to select
+     *                 documents.
+     * @see <a
+     * href="https://console.bluemix.net/docs/services/Cloudant/api/cloudant_query.html#selector-syntax"
+     * target="_blank">selector syntax</a>
+     * @see com.cloudant.client.api.Database#query(String, Class)
+     * @see Selector
+     */
+    public QueryBuilder(String selector) {
```

Comment:
I'm leaning back towards using the `Selector` type here, given that `Database#query(..)` accepts a `String` for the body there is still a way to workaround any selector limitations without having to accept the `String` here.

Reply:
Done in b172759